### PR TITLE
ESQL: Begin moving LogicalPlan to NamedWriteable

### PR DESCRIPTION
**Pull request by nik9000**
Now all `Node` subclasses are `NamedWriteable`! Except the `PhysicalPlan`s don't implement it properly yet. And only a hand full of `LogicalPlan`s do. But it's more progress!

We now see a light at the end of the tunnel for migrating serialization: there's just about a screen worth of `LogicalPlan` subclasses to migrate and two screens worth of `PhysicalPlan`s.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java [hunk 1]
```diff
@@ -44,8 +44,8 @@ public static JoinType readFrom(StreamInput in) throws IOException {
             case 0 -> INNER;
             case 1 -> LEFT;
             case 2 -> RIGHT;
-            case 4 -> FULL;
-            case 5 -> CROSS;
+            case 3 -> FULL;
+            case 4 -> CROSS;
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java:48]: Good catch. 
But should this not be a bug fix? 8.15 will have the wrong IDs (there should be one more BC for it, so in time).

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/AbstractNodeSerializationTests.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+public abstract class AbstractNodeSerializationTests<T extends Node<? super T>> extends AbstractWireTestCase<T> {
+    /**
+     * We use a single random config for all serialization because it's pretty
+     * heavy to build, especially in {@link #testConcurrentSerialization()}.
+     */
+    private EsqlConfiguration config;
+
+    public static Source randomSource() {
+        int lineNumber = between(0, EXAMPLE_QUERY.length - 1);
+        int offset = between(0, EXAMPLE_QUERY[lineNumber].length() - 2);
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/AbstractNodeSerializationTests.java:40]: nit/optional: maybe extract a `String line = EXAMPLE_QUERY[lineNumber];`?

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/logical/LookupSerializationTests.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        switch (between(0, 3)) {
+            case 0 -> child = randomValueOtherThan(child, () -> randomChild(0));
+            case 1 -> tableName = randomValueOtherThan(tableName, AbstractExpressionSerializationTests::randomChild);
+            case 2 -> matchFields = randomValueOtherThan(matchFields, () -> randomFieldAttributes(1, 10, false));
+            case 3 -> localRelation = randomValueOtherThan(
+                localRelation,
+                () -> randomBoolean() ? null : LocalRelationSerialiationTests.randomLocalRelation()
+            );
+        }
+        return new Lookup(source, child, tableName, matchFields, localRelation);
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/logical/LookupSerializationTests.java:51]: Should the input source be retained?

**Review by bpintea — APPROVED** [review]
Lgtm

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java:48]: The test really caught it. In production we only use LEFT at the moment. Have for a while. It's less exciting than it looks.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/logical/LookupSerializationTests.java (hunk 3 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/logical/LookupSerializationTests.java:51]: :+1:

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/AbstractNodeSerializationTests.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/AbstractNodeSerializationTests.java:40]: Yeah. Seems more readable.

**Pull request merged**